### PR TITLE
[io] add support for wildcards in File RecordStream

### PR DESCRIPTION
**Pull request by luca-s**
File::addStream now supports * ? () |

This is related to https://github.com/SeisComP/main/pull/59 - without wildcards support that PR is of limited use.

### CODE REVIEW

**Comment by gempa-jabe** [comment]
This will break the ABI and could only be used with SC >= 6. I have had decided against stream ID filtering in the file RS for performance reasons. Furthermore full regexp support is not in SDS archive and I would prefer to keep both filter options in sync. That is why `wildcmp` should be used rather than regexp. The current implementation only filters records if the streamID is wildcarded otherwise no filter is in place. If that is correct, does that make sense? Adding `XX.ABCD..BHZ` would not have an effect?

If all that is really required to make `scart` work we need to carefully design the file rs to not break existing code.

**Comment by luca-s** [comment]
> This will break the ABI and could only be used with SC >= 6. 

I am aware of it and it is not an issue.

>  I have had decided against stream ID filtering in the file RS for performance reasons. 

I thought about it while modifying the code and this PR doesn't slow down the previous behaviour of not having wildcards. So the performance hit, if there is a hit, can be seen only if the wildcards are in use. But do you really think there could be a problem with performance? I don't see a real world use case where this can be a problem, but I can see a  couple of scenarios where this feature might be useful. However I don't want to insist. Just some thoughts.

>  Furthermore full regexp support is not in SDS archive and I would prefer to keep both filter options in sync. That is why wildcmp should be used rather than regexp. 

Good point I can change it. I don't believe loosing the ability to make use of ( ) and | is a big deal. 

> The current implementation only filters records if the streamID is wildcarded otherwise no filter is in place. 

Are you talking about wildcmp or File rs now?


> If all that is really required to make scart work we need to carefully design the file rs to not break existing code.

This change is not required, `scart` can work without it, but the user would only be able to filter with fully qualified stream codes, that is without wildcards. However, since `scart` was never able to use stream/time filtering with File as input, then I don't see any problem with that.

**Comment by gempa-jabe** [comment]
> Good point I can change it. I don't believe loosing the ability to make use of ( ) and | is a big deal.

And if we need it anytime in future it could be added to `wildcmp`.

> Are you talking about wildcmp or File rs now?

The file rs. I see that I did a mistake interpreting the code. I think using `wildcmp` will be fine and if carefully implemented will not hit the performance as you said.

**Comment by gempa-dirk** [comment]
> This change is not required, `scart` can work without it, but the user would only be able to filter with fully qualified stream codes, that is without wildcards. However, since `scart` was never able to use stream/time filtering with File as input, then I don't see any problem with that.

I found the stream and time filtering useful when generating SDS from files and I would be happy if this new feature could be kept. Otherwise, the intermediate step via scmssort is required but of course doable.

File: libs/seiscomp/io/recordstream/file.cpp [hunk 1]
```diff
@@ -211,13 +221,44 @@ bool File::setRecordType(const char *type) {
...
+	const auto & it = _filter.find(rec->streamID());
+	if ( it != _filter.end() ) {
+		return &it->second;
+	}
+
+	// then search the wildcarded filters
+	for ( const auto& pair : _reFilter ) {
+		const string& wild = pair.first;
+		const TimeWindowFilter& twf = pair.second;
+		if ( Core::wildcmp(wild, rec->streamID()) ) {
```
Comment (gempa-jabe) [inline libs/seiscomp/io/recordstream/file.cpp:237]: `rec->streamID()` is rather expensive as it composes the string on demand. Please save this value at the top of the function. It should be called only once.

**Comment by gempa-jabe** [comment]
The code is looking good to me and I would merge it. Any blockers?

**Comment by gempa-jabe** [comment]
I have tested it and it worked. Thanks!

**Pull request merged**